consensus/tendermint/tendermint.go

```diff
@@ -0,0 +1,637 @@
+package tendermint
+
+import (
+	"fmt"
+	"maps"
+	"slices"
+	"sync"
+	"time"
+
+	"github.com/NethermindEth/juno/core/felt"
+)
+
+type step uint
+
+const (
+	propose step = iota
+	prevote
+	precommit
+)
+
+type timeoutFn func(round uint) time.Duration
+
+type Addr interface {
+	// Ethereum Addresses are 20 bytes
+	~[20]byte | felt.Felt
+}
+
+type Hash interface {
+	~[32]byte | felt.Felt
+}
+
+// Hashable's Hash() is used as ID()
+type Hashable[H Hash] interface {
+	Hash() H
+}
+
+type Application[V Hashable[H], H Hash] interface {
+	// Value returns the value to the Tendermint consensus algorith which can be proposed to other validators.
+	Value() V
+
+	// Valid returns true if the provided value is valid according to the application context.
+	Valid(V) bool
+}
+
+type Blockchain[V Hashable[H], H Hash, A Addr] interface {
+	// Height return the current blockchain height
+	Height() uint
+
+	// Commit is called by Tendermint when a block has been decided on and can be committed to the DB.
+	Commit(uint, V, []Precommit[H, A])
+}
+
+type Validators[A Addr] interface {
+	// TotalVotingPower represents N which is required to calculate the thresholds.
+	TotalVotingPower(height uint) uint
+
+	// ValidatorVotingPower returns the voting power of the a single validator. This is also required to implement
+	// various thresholds. The assumption is that a single validator cannot have voting power more than f.
+	ValidatorVotingPower(validatorAddr A) uint
+
+	// Proposer returns the proposer of the current round and height.
+	Proposer(height, round uint) A
+}
+
+type Slasher[M Message[V, H, A], V Hashable[H], H Hash, A Addr] interface {
```

Comment:
Any validator (or entity that has the relevant messages etc) can call the L2 contract?

Reply:
Yes, note you don't need a quorum of validators because it is assumed that signatures cannot be forged. Therefore, even one validator sending 2 equivocated messages is sufficient for slashing.